### PR TITLE
MULE-10339: Replace Function<MuleEvent, T> with a better abstraction

### PR DESCRIPTION
**Pull request by estebanwasinger**
Requires:
https://github.com/mulesoft/mule-extensions-api/pull/170
https://github.com/mulesoft/metadata-model-api/pull/33

### CODE REVIEW

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/introspection/describer/AnnotationsBasedDescriber.java [hunk 1]
```diff
@@ -584,14 +589,27 @@ private void addInterceptingCallbackModelProperty(MethodElement operationMethod,
...
+  }
+
+  private class ParameterResolverParameterTypeContributor implements ParameterDeclarerContributor {
+
+    @Override
+    public void contribute(ExtensionParameter parameter, ParameterDeclarer declarer) {
+      MetadataType metadataType = parameter.getMetadataType(typeLoader);
+      if (ParameterResolver.class.isAssignableFrom(parameter.getType().getDeclaringClass())) {
+        final Optional<MetadataType> expressionResolverType = getGenericTypeAt(metadataType, 0, typeLoader);
+        metadataType = expressionResolverType.isPresent() ? expressionResolverType.get() : typeLoader.load(Object.class);
```
Comment (marianogonzalez) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/introspection/describer/AnnotationsBasedDescriber.java]: this should throw IllegalOperationModelDefinitionException

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/resolver/ParameterResolverArgumentResolver.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    argumentResolver = new ByParameterNameArgumentResolver<>(parameterName);
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public ParameterResolver<T> resolve(OperationContext operationContext) {
+    Object value = argumentResolver.resolve(operationContext);
+    return value != null && ParameterResolver.class.isAssignableFrom(value.getClass()) ? (ParameterResolver<T>) value
```
Comment (marianogonzalez) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/resolver/ParameterResolverArgumentResolver.java]: ParameterResolver.class.isInstance(value)

Comment (marianogonzalez) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/resolver/ParameterResolverArgumentResolver.java]: put each ternary segment on its own line

File: modules/extensions-spring-support/src/main/java/org/mule/runtime/module/extension/internal/config/dsl/ExtensionDefinitionParser.java [hunk 3]
```diff
@@ -699,6 +716,13 @@ private boolean isExpressionFunction(MetadataType metadataType) {
     }
   }
 
+  private boolean isExpressionResolver(Set<ModelProperty> modelProperties) {
+    return modelProperties
+        .stream()
+        .filter(property -> property instanceof ParameterResolverTypeModelProperty)
+        .findAny()
+        .isPresent();
```
Comment (alegmarra) [inline modules/extensions-spring-support/src/main/java/org/mule/runtime/module/extension/internal/config/dsl/ExtensionDefinitionParser.java]: anyMatch

**Pull request merged**